File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
     
        vector<int> v (2,-1);
        int low = lower( nums, target);
        if(low == -1)
            return v;
        int upp = upper( nums, target);
            
        v[0]=low;
        if(upp==-1)
            upp=nums.size();
        v[1]=upp-1;
        return v;
        
    }
    
    int lower(vector<int>& nums, int target) {
        int res =-1;
       if(nums.size()==0||target < nums[0] || target > nums[nums.size()-1])
            return res;
        int left = 0 ; 
        int right = nums.size()-1;
        
        while(left<=right)
        {
            int mid= left+(right-left)/2;
            if(nums[mid] == target)
            {
                 res = mid;
                right = mid -1 ;            
            }
            else if (target < nums[mid])
                    right = mid -1 ;
            else
                left=mid+1;
        }
        
        return res;
    }
    int upper(vector<int>& nums, int target) {
        int left = 0 ; 
        int res =-1;
        int right = nums.size()-1;
        
        while(left<=right)
        {
            int mid= left+(right-left)/2;
            if(nums[mid] > target)
            {
                res = mid;
                right = mid-1;
            }
            else
                left=mid+1;
            
        }
        
        return res;
    }

};

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {

    // Main method to return the first and last positions of the target in the sorted array
    public int[] searchRange(int[] nums, int target) {
        // Perform left-biased binary search to find the first occurrence
        int leftMost = binarySearch(nums, target, true);
        // Perform right-biased binary search to find the last occurrence
        int rightMost = binarySearch(nums, target, false);

        // Return both indices as an array
        return new int[] { leftMost, rightMost };
    }

    /**
     * Performs a modified binary search to find either the first or last occurrence of the target.
     *
     * @param nums       Sorted input array
     * @param target     Value to search for
     * @param leftBiased If true, find first occurrence; else, find last occurrence
     * @return           Index of the occurrence, or -1 if not found
     *
     * ⏱ Time Complexity: O(log n)
     *     - Standard binary search on sorted array takes log n time.
     *     - This function is called twice, but total remains O(log n).
     *
     * \U0001f9e0 Space Complexity: O(1)
     *     - Uses only constant extra space (no recursion or data structures).
     */
    private int binarySearch(int[] nums, int target, boolean leftBiased) {
        int result = -1;                // Default if target not found
        int left = 0;
        int right = nums.length - 1;

        while (left <= right) {
            // Prevent potential overflow compared to (left + right) / 2
            int mid = left + (right - left) / 2;

            if (target < nums[mid]) {
                // Target is in the left half
                right = mid - 1;
            } else if (target > nums[mid]) {
                // Target is in the right half
                left = mid + 1;
            } else {
                // Found a match
                result = mid;

                if (leftBiased) {
                    // Keep searching left side for first occurrence
                    right = mid - 1;
                } else {
                    // Keep searching right side for last occurrence
                    left = mid + 1;
                }
            }
        }

        return result;
    }
}
